### PR TITLE
add support for DER and P12 certs (#2411)

### LINKED ISSUE #2411: Add support for P12 cert format via CURLOPT_SSLCERTTYPE
**Description**  
New version of openssl (versions 0.9.3 and later) suport P12 certificate files, it would be nice to be able to specify the type as the third parameter of `cert` or even detect them by the extension maybe and set the correct CURLOPT_SSLCERTTYPE value. Here is the official documentation: https://curl.haxx.se/libcurl/c/CURLOPT_SSLCERTTYPE.html

**Example**  
it currently works with specifying the `curl` option manually:
```
$response = $client->post($url, [
    RequestOptions::HEADERS => ['apns-topic' => $topic],
    RequestOptions::JSON => json_decode('{}'),
    RequestOptions::VERSION => 2.0,
    RequestOptions::CERT => ['./certificates/cert.p12', '123456'],
    'curl' => [
        CURLOPT_SSLCERTTYPE => 'P12'
    ]
]);
```

I can prepare a PR if this change is welcome. Thank you!

### PR DESCRIPTION
**Pull request by razvanphp**
https://curl.haxx.se/libcurl/c/CURLOPT_SSLCERTTYPE.html

This will fix #2411

### CODE REVIEW

**Comment by Nyholm** [comment]
I think this makes sense. Could you add a test for it?

Also, it needs a rebase.

**Comment by Nyholm** [comment]
Friendly ping

**Comment by razvanphp** [comment]
ok, I will try to do that.

**Comment by GrahamCampbell** [comment]
I think this could be added to the 7.3.0 milestone (for November??? release).

**Comment by maksimovic** [comment]
Any idea when this will make its way to the release? We rely on Guzzle and are in need of P12 cert support pretty urgently, so I'm checking in here to see what our options are. Thanks.

**Comment by Nyholm** [comment]
I need a test for this PR before we can review and merge it. 
After merge we are usually pretty quick to release it. 

We haven't heard from the author for about a year. @maksimovic feel free to open a new PR with this fix and a test. That will probably be the quickest way to get this released.

**Comment by Nyholm** [comment]
Ooops. @razvanphp answered a month ago.. What is the status? Do you need help from @maksimovic?

**Comment by razvanphp** [comment]
I'm on it 🙂 didn't work on the project that needed this for some time...

**Comment by razvanphp** [comment]
Rebased, tests added, thank you!

**Comment by Nyholm** [comment]
Thank you

@maksimovic Could you give this a review and make sure that it works?

File: tests/Handler/CurlFactoryTest.php
Comment (Nyholm) [inline tests/Handler/CurlFactoryTest.php:265]: We dont want to fail the test because we cannot clean up after ourselves. 


```suggestion
        $certFile = tempnam(sys_get_temp_dir(), "mock_test_cert");
        rename($certFile, $certFile .= '.der');
        try {
            $f = new Handler\CurlFactory(3);
            $f->create(new Psr7\Request('GET', Server::$url), ['cert' => $certFile]);
            self::assertArrayHasKey(\CURLOPT_SSLCERTTYPE, $_SERVER['_curl']);
            self::assertEquals('DER', $_SERVER['_curl'][\CURLOPT_SSLCERTTYPE]);
        } finally {
            \@unlink($certFile);
        }
```

Comment (razvanphp) [reply tests/Handler/CurlFactoryTest.php:265]: true, but I just followed the model from `StreamHandlerTest`

Comment (razvanphp) [reply tests/Handler/CurlFactoryTest.php]: ... but I've pushed the suggested change nevertheless, let's get this merged :)

**Comment by razvanphp** [comment]
@maksimovic ping :)

**Comment by maksimovic** [comment]
Unfortunately, we opted out of using Guzzle because we're using an older version it, but can't upgrade it quickly since it's the vital part of our stack and also includes bringing some other dependencies to a more recent version, which then have their dependencies requiring version bump, so it turned out to be way more work that we anticipated.

**Comment by razvanphp** [comment]
@Nyholm what do you think? can we merge this tho? I've already tested it in our application, also the tests are included in the PR.

**Comment by nuernbergerA** [comment]
Hey @Nyholm we need p12 support in our application too.

Is there still anything blocking this from being merged?

**Comment by Nyholm** [comment]
@nuernbergerA Have you tested this patch in your app? Does it work as expected?

**Comment by nuernbergerA** [comment]
@Nyholm yes it works.

But if there is any reason you want to reject the PR I'm fine to include the `curl` key.
```php
'curl' => [
    CURLOPT_SSLCERTTYPE => 'P12',
],
```

**Comment by Nyholm** [comment]
No. I don't want to reject it. I want to be comfortable that merging this fixes the issue and does not break anything. 

I'm not familiar with the feature but since many people request this, I encourage people to verify the patch and not only bump the thread.

**Review by Nyholm — APPROVED** [review]
Thank you

**Pull request merged**
